localstack/packages/__init__.py

```diff
@@ -0,0 +1,349 @@
+
+    pass
+
+
+class NoSuchVersionException(PackageException):
+    """Exception indicating that a requested installer version is not available / supported."""
+
+    pass
+
+
+class SystemNotSupportedException(PackageException):
+    """Exception indicating that the current system is not allowed."""
+
+    pass
+
+
+class InstallTarget(Enum):
+    """
+    Different installation targets.
+    Attention:
+    - These targets are directly used in the LPM API and are therefore part of a public API!
+    - The order of the entries in the enum define the default lookup order when looking for package installations.
+
+    These targets refer to the directories in config#Directories.
+    - VAR_LIBS: Used for packages installed at runtime. They are installed in a host-mounted volume.
+                This directory / these installations persist across multiple containers.
+    - STATIC_LIBS: Used for packages installed at build time. They are installed in a non-host-mounted volume.
+                   This directory is re-created whenever a container is recreated.
+    """
+
+    VAR_LIBS = config.dirs.var_libs
+    STATIC_LIBS = config.dirs.static_libs
+
+
+class PackageInstaller(abc.ABC):
+    """
+    Base class for a specific installer.
+    An instance of an installer manages the installation of a specific Package (in a specific version, if there are
+    multiple versions).
+    """
+
+    def __init__(self, name: str, version: str):
+        """
+        :param name: technical package name, f.e. "opensearch"
+        :param version: version of the package to install
+        """
+        self.name = name
+        self.version = version
+
+    def install(self, target: Optional[InstallTarget] = None) -> None:
+        """
+        Performs the package installation.
+
+        :param target: preferred installation target. Default is VAR_LIBS.
+        :return: None
+        :raises PackageException: if the installation fails
+        """
+        try:
+            if not target:
+                target = InstallTarget.VAR_LIBS
+            if not self.is_installed():
+                LOG.debug("Starting installation of %s...", self.name)
+                self._prepare_installation(target)
+                self._install(target)
+                self._post_process(target)
+                LOG.debug("Installation of %s finished.", self.name)
+            else:
+                LOG.debug("Installation of %s skipped (already installed).", self.name)
+        except PackageException as e:
+            raise e
+        except Exception as e:
+            raise PackageException(f"Installation of {self.name} failed.") from e
+
+    def is_installed(self) -> bool:
+        """
+        Checks if the package is already installed.
+
+        :return: True if the package is already installed (i.e. an installation is not necessary).
+        """
+        return self.get_installed_dir() is not None
```

Comment:
couldn't it be the case that the install directory was created but the installation never proceeded? it would seem to me to be better to check whether some key files (like executables) exist.

Reply:
I agree, thanks for the hint! I introduced a function to the interface which builds an "install marker" based on the given installation directory. This is a bit more abstract than the "executable path" discussed below (but for the current implementations is essentially the same).

Reply:
👍